src/main/java/com/github/manolo8/darkbot/core/objects/facades/AstralGateProxy.java

```diff
@@ -1,21 +1,137 @@
 package com.github.manolo8.darkbot.core.objects.facades;
 
 import com.github.manolo8.darkbot.core.itf.Updatable;
+import com.github.manolo8.darkbot.core.objects.swf.ObjArray;
 import com.github.manolo8.darkbot.core.utils.ByteUtils;
 
 import static com.github.manolo8.darkbot.Main.API;
+import static com.github.manolo8.darkbot.Main.UPDATE_LOCKER;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public class AstralGateProxy extends Updatable {
-    public int highScore, currentRift, currentScore, cpuCount;
+    private int highScore, currentRift, currentScore, cpuCount;
+    private boolean canEquip;
+    private List<AstralItem> rewardItems = new ArrayList<>();
+    private List<AstralItem> inventoryItems = new ArrayList<>();
+
+    private final ObjArray rewardItemsArr = ObjArray.ofVector(true);
+    private final ObjArray inventoryItemsArr = ObjArray.ofVector(true);
 
     @Override
     public void update() {
+        if (address == 0) {
+            return;
+        }
+
         long data = API.readMemoryLong(address + 48) & ByteUtils.ATOM_MASK;
 
-        highScore = API.readMemoryInt(data + 64);
-        currentRift = API.readMemoryInt(API.readMemoryLong(data + 80) + 40);
-        currentScore = API.readMemoryInt(API.readMemoryLong(data + 88) + 40);
-        cpuCount = API.readMemoryInt(API.readMemoryLong(data + 96) + 40);
+        this.highScore = API.readMemoryInt(data + 64);
+        this.currentRift = API.readMemoryInt(API.readMemoryLong(data + 80) + 40);
+        this.currentScore = API.readMemoryInt(API.readMemoryLong(data + 88) + 40);
+        this.cpuCount = API.readMemoryInt(API.readMemoryLong(data + 96) + 40);
+        this.canEquip = API.readBoolean(API.readMemoryLong(data + 0x0B0) + 0x20);
+
+        rewardItemsArr.update(API.readMemoryLong(data + 0x88));
+        synchronized (UPDATE_LOCKER) {
+            rewardItemsArr.sync(rewardItems, AstralItem::new);
+        }
+
+        inventoryItemsArr.update(API.readMemoryLong(data + 0x0A0));
+        synchronized (UPDATE_LOCKER) {
```

Comment:
those update locks can probably go away too

Reply:
the update locker exists for the **bot** gui not to re-paint while stuff is being updated, NOT the in-game gui.

If the data is not being painted by darkbot's MapDrawer, this serves no purpose